### PR TITLE
Fix rotation and nslayout line drawing

### LINKED ISSUE #1275: Rotation broken for drawing text
My most recent math changes causes text to rotate incorrectly when drawing.

### PR DESCRIPTION
**Pull request by aballway**
Fixes incorrect rotation introduced by my recent math changes and adds descent to nslayoutmanager line drawing to make exclusion zones more accurate

Fixes #1275

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1274)

### CODE REVIEW

File: Frameworks/UIKit/NSLayoutManager.mm [hunk 1]
```diff
@@ -247,9 +247,9 @@ - (void)drawGlyphsForGlyphRange:(NSRange)range atPoint:(CGPoint)position {
         CTLineRef line = (CTLineRef)_ctLines[curLine];
         CGContextSaveGState(curCtx);
 
-        CGFloat ascent, leading;
+        CGFloat ascent, descent, leading;
         CTLineGetTypographicBounds(line, &ascent, nullptr, &leading);
```
Comment (ms-jihua) [inline Frameworks/UIKit/NSLayoutManager.mm]: (i think ascent + leading would've been correct here in any case)

---
In reply to: [85786676](https://github.com/Microsoft/WinObjC/pull/1274#discussion_r85786676) [](ancestors = 85786676)

File: Frameworks/CoreGraphics/CGContextCairo.mm [hunk 2]
```diff
@@ -1868,18 +1868,19 @@
...
     CGAffineTransform userTransform =
         CGAffineTransformConcat(curState->curTransform, CGAffineTransformMake(1.0f / _scale, 0, 0, 1.0f / _scale, 0, -height / _scale));
 
-    // Apply the context CTM
+    // Apply the context CTM to get total matrix
     transform = CGAffineTransformConcat(transform, userTransform);
 
     // Perform anti-clockwise rotation required to match the reference platform.
-    imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, -transform.b, transform.c, transform.d, transform.tx, -transform.ty));
+    imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, -transform.b, -transform.c, transform.d, transform.tx, -transform.ty));
```
Comment (ms-jihua) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: >- [](start = 78, length = 1)

i'd been wondering for a while why this was missing...

**Comment by ms-jihua** [comment]
:shipit:

File: tests/testapps/CTCatalog/CTCatalog.vsimporter/CTCatalog-WinStore10/CTCatalog.vcxproj [hunk 3]
```diff
@@ -177,6 +177,7 @@
     <AppxManifest Include="Package.appxmanifest">
       <SubType>Designer</SubType>
     </AppxManifest>
+    <ClangCompile Include="..\..\CTCatalog\Tests\CTCAffineTransformationTestViewController.mm" />
```
Comment (rajsesh-msft) [inline tests/testapps/CTCatalog/CTCatalog.vsimporter/CTCatalog-WinStore10/CTCatalog.vcxproj]: These files also need to be added to xcodeproject and should compile and run on reference platform.

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    [_scaleYSlider setNeedsDisplay];
+
+    _translateXSlider.frame = CGRectMake(2.0f * width / 3, 10, width / 4, 50);
+    [_translateXSlider setNeedsDisplay];
+
+    _translateYSlider.frame = CGRectMake(2.0f * width / 3, 70, width / 4, 50);
+    [_translateYSlider setNeedsDisplay];
+}
+
+- (void)didReceiveMemoryWarning {
```
Comment (rajsesh-msft) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm]: these functions can be deleted.

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//*****************************************************************************
+
+#import "CTCAffineTransformationTestViewController.h"
+
+static const NSString* text = @"The quick brown dog jumps over the lazy fox. THE QUICK BROWN DOG JUMPS OVER THE LAZY FOX.";
```
Comment (rajsesh-msft) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm]: I would change this text to read what the expected test result should be (changing rotate should cause the text to rotate in which direction?).

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    // Creates path with current rectangle
+    CGMutablePathRef path = CGPathCreateMutable();
+    CGPathAddRect(path, NULL, self.bounds);
+
+    // Create style setting to match given alignment
+    CTParagraphStyleSetting setting;
+
+    setting.spec = kCTParagraphStyleSpecifierAlignment;
+    setting.valueSize = sizeof(CTTextAlignment);
+    CTTextAlignment alignment = kCTCenterTextAlignment;
```
Comment (rajsesh-msft) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm]: left alignment here to easily see the pivot point?

File: tests/testapps/CTCatalog/CTCatalog.vsimporter/CTCatalog-WinStore10/CTCatalog.vcxproj (hunk 3 above)
Comment (aballway) [inline tests/testapps/CTCatalog/CTCatalog.vsimporter/CTCatalog-WinStore10/CTCatalog.vcxproj]: Already done

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    // Creates frame for framesetter with current attributed string
+    CTFrameRef frame = CTFramesetterCreateFrame(framesetter, CFRangeMake(0, 0), path, NULL);
+    CFAutorelease(frame);
+
+    // Draws the text in the frame
+    CTFrameDraw(frame, context);
+
+    // Creates outline
+    CGContextSetLineWidth(context, 2.0);
+    CGColorSpaceRef colorspace = CGColorSpaceCreateDeviceRGB();
```
Comment (DHowett-MSFT) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCAffineTransformationTestViewController.mm]: This colorspace does not seem to be used.

**Pull request merged**